### DISCUSSION TITLE
可以根据不同架构切换使用不同的Qt路径吗

### DISCUSSION THREAD

**Discussion by Elite-stay** [xmake-io/xmake#5294] [Q&A] [answered]
指windows下x86和x64的路径切换，自己写个包描述可行？但是不知道该怎么写

**Comment by star-hengxing** [comment]
试试

```
$ xmake f -a x64 --qt_path=path/to/qt
```

```lua
option("qt_path", default = nil)

if has_config("qt_path") then
    set_config("qt", path.join(get_config("qt_path"), get_config("arch")))
end
```

**Reply by Elite-stay** [reply]
这样只能一个个工程去写吧，我试着写个package不太行，rule的话还不知道怎么集中放在一起引入

```lua
package("myqt")
    set_description("myqt")

    on_load(function (package)
        -- root
        package:set("installdir", path.join(os.scriptdir(), "../../../.."))
    end)

    on_fetch(function (package)
        local arch = package:arch()
        if arch == "x86" then
            package:config_set("qt", package:installdir("myqt/msvc2013_opengl"))
        else
            package:config_set("qt", package:installdir("myqt/msvc2013_64_opengl"))
        end
        return {}
    end)
```

**Comment by Elite-stay** [comment]
我又试了好多方法，感觉还是卡住了，我希望的是，实际工程只需要设置一个公共目录的 option ，就可以根据 option 来导入一个 rule ，这个 rule 里面根据架构判断使用公共目录的 x64 的 qt 还是 x86 的 qt ，比如是这样的

```lua
-- myqt.lua
rule("myqt")
    on_load(function (target)
        if target:is_arch("x86") then
            target:set_config("qt", path.join("$(Public)", "msvc2013_opengl"))
        else
            target:set_config("qt", path.join("$(Public)", "msvc2013_64_opengl"))
        end
    end)
rule_end()
```

```lua
-- xmake.lua
option("Public")
    set_default("$(env Public)")
option_end()

includes("$(Public)/rules/myqt.lua")
add_rules("myqt")
```

但是 includes 不识别`$(Public)`，不知道可以怎么做了

**Reply by waruqi** [reply]
不要在 includes 里面这么搞，includes 导入 lua 脚本后，才能解析 option ，你又在 includes 阶段就要完成 option 解析，这不是鸡生蛋问题么。。

includes 应该无脑设置，不管啥状态，你先 includes 全部引入进来，干嘛要分 option 状态。。

option 状态的判断是 rule 里面做的。。

> target:set_config("qt", path.join("$(Public)", "msvc2013_opengl"))

另外，没 target:set_config 这种接口，不要瞎搞。。也不要在脚本域使用 `$(var)` 内置变量。。大部分接口都不支持的，这玩意通常用于描述域。。获取 option 配置 用 `get_config`  。

而 set_config 是全局描述域接口，并且需要在最开头，最起始位置设置。。脚本域不要用。。

**Reply by waruqi** [reply, marked as answer]
你可以尝试重新定义 qt.env rule 来重写查找规则。。不过就是有些 hack ，以后有可能会被 break

https://github.com/xmake-io/xmake/blob/dev/xmake/rules/qt/env/xmake.lua

或者参考 xmake-repo 的 qt 包，qtbase,qtgui ,...  做些定制

https://github.com/xmake-io/xmake-repo/tree/dev/packages/q/qtbase

**Reply by Elite-stay** [reply]
> 你可以尝试重新定义 qt.env rule 来重写查找规则。。不过就是有些 hack ，以后有可能会被 break
> 
> https://github.com/xmake-io/xmake/blob/dev/xmake/rules/qt/env/xmake.lua
> 
> 或者参考 xmake-repo 的 qt 包，qtbase,qtgui ,... 做些定制
> 
> https://github.com/xmake-io/xmake-repo/tree/dev/packages/q/qtbase

qtbase 包的 on_fetch 好像有点启发，明天试试，package 主要是 local 包文档不是特别详细，不知道 on_fetch 可以返回哪些内容

**Comment by Elite-stay** [comment]
ok可以了，写了个包描述
